File: java/sieve/src/main/java/Sieve.java
import java.util.ArrayList;
import java.util.List;

class Sieve {
    private final int maxPrime;

    Sieve(int maxPrime) {
        this.maxPrime = maxPrime;
    }

    List<Integer> getPrimes() {
        final var primes = new ArrayList<Integer>();

        for (var i = 2; i <= maxPrime; i++) {
            if (i == 2) {
                primes.add(i);
                continue;
            }

            if (i % 2 == 0) {
                continue;
            }

            var hasDivisor = false;
            for (var x = 3; x <= Math.sqrt(i); x++) {
                if (i % x == 0) {
                    hasDivisor = true;
                    break;
                }
            }

            if (!hasDivisor) {
                primes.add(i);
            }
        }

        return primes;
    }
}

File: kotlin/sieve/src/main/kotlin/Sieve.kt
import kotlin.math.sqrt

object Sieve {

    fun primesUpTo(upperBound: Int): List<Int> {
        return (2..upperBound).filter {
            if (it == 2) return@filter true
            if (it % 2 == 0) return@filter false

            for (x in 3..(sqrt(it.toDouble()).toInt())) {
                if (it % x == 0) return@filter false
            }

            true
        }
    }
}
